packages/@lwc/engine-dom/src/renderer-factory.ts

```diff
@@ -0,0 +1,352 @@
+/*
+ * Copyright (c) 2018, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+
+/**
+ * Note: This module cannot import any modules because it is meant to be a pure function.
+ *   Type-Only imports are allowed.
+ */
+import type { RendererAPI } from '@lwc/engine-core';
+
+/**
+ * A factory function that produces a renderer.
+ * Renderer encapsulates operations that are required to render an LWC component into the underlying
+ * runtime environment. In the case of @lwc/enigne-dom, it is meant to be used in a DOM environment.
+ * @param createFragment
+ * @param globalThis
+ * @param insertStylesheet
+ * @param htmlPropertyToAttribute
+ * @returns {@link RendererAPI}
+ */
+export function rendererFactory(
+    createFragment: (html: string) => Node | null,
+    globalThis: any,
+    insertStylesheet: (content: string, target?: ShadowRoot) => void,
+    htmlPropertyToAttribute: (propName: string) => string
+): RendererAPI {
+    // Util functions
+    function assertFail(msg: string): never {
+        throw new Error(msg);
+    }
+    function assertInvariant(value: any, msg: string): asserts value {
+        if (!value) {
+            throw new Error(`Invariant Violation: ${msg}`);
+        }
+    }
+
+    function isNull(obj: any): obj is null {
+        return obj === null;
+    }
+
+    function isUndefined(obj: any): obj is undefined {
+        return obj === undefined;
+    }
+
+    // Note: These keys have to be kept in sync with @lwc/shared
+    const KEY__IS_NATIVE_SHADOW_ROOT_DEFINED = '$isNativeShadowRootDefined$';
+    const KEY__SHADOW_TOKEN = '$shadowToken$';
```

Comment:
Why duplicate these properties from `@lwc/shared`?

Reply:
The renderer-factory needs to be a pure function without any outside dependencies(imports). The goal is that it will be toString-ed and eval-ed inside a sandbox. Invocations the sandboxed renderer-factory will produce a renderer.